### PR TITLE
[BEAM-1984] Solve issues with runtime/test dependencies not been correctly scoped

### PR DESCRIPTION
**Pull request by iemejia**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by iemejia** [comment]
R: @aviemzur

**Comment by iemejia** [comment]
Retest this please

**Comment by iemejia** [comment]
Retest this please

**Comment by iemejia** [comment]
Run Spark RunnableOnService

**Comment by iemejia** [comment]
Run Dataflow ValidatesRunner

**Comment by iemejia** [comment]
Run Spark ValidatesRunner

**Comment by iemejia** [comment]
Run Dataflow ValidatesRunner

**Comment by iemejia** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
R: @dhalperi for the changes in IOs.

**Comment by iemejia** [comment]
Ok this must be ok now. Please review guys and give me your comments or LGTM.

File: pom.xml [hunk 1]
```diff
@@ -1278,15 +1285,15 @@
...
+              <version>2.20</version>
             </dependency>
           </dependencies>
         </plugin>
 
         <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-failsafe-plugin</artifactId>
-          <version>2.19.1</version>
+          <version>2.20</version>
```
Comment (dhalperi) [inline pom.xml]: Why is this needed in this PR? If not, please upgrade deps in separate PR from refactoring.

File: runners/spark/pom.xml [hunk 2]
```diff
@@ -200,9 +201,8 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
-      <groupId>commons-lang</groupId>
-      <artifactId>commons-lang</artifactId>
-      <version>2.6</version>
```
Comment (dhalperi) [inline runners/spark/pom.xml]: What does "align" mean here -- align with what? Is the claim that Spark already depends on `lang3` so we should just use it, e.g.?

How does `mvn dependency:list` change with this switch?

File: sdks/java/extensions/sorter/pom.xml [hunk 3]
```diff
@@ -28,33 +28,11 @@
 
   <artifactId>beam-sdks-java-extensions-sorter</artifactId>
   <name>Apache Beam :: SDKs :: Java :: Extensions :: Sorter</name>
-  
+
   <properties>
-    <hadoop.version>2.7.1</hadoop.version>
+    <hadoop.version>2.7.3</hadoop.version>
```
Comment (dhalperi) [inline sdks/java/extensions/sorter/pom.xml]: why?

File: sdks/java/io/common/pom.xml [hunk 4]
```diff
@@ -33,10 +33,12 @@
       <dependency>
         <groupId>org.apache.beam</groupId>
         <artifactId>beam-sdks-java-core</artifactId>
+        <scope>test</scope>
```
Comment (dhalperi) [inline sdks/java/io/common/pom.xml]: Hmm. correct, but odd that this is correct. I'd expect `HashingFn` to be in `main`, not `test` in that module. Irrelevant to this PR, but CC: @ssisk

File: pom.xml [hunk 5]
```diff
@@ -1267,7 +1267,7 @@
         <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-surefire-plugin</artifactId>
-          <version>2.19.1</version>
+          <version>2.20</version>
```
Comment (dhalperi) [inline pom.xml]: please extract a property (`surefire.version`?) and move 2.20 into it :)

File: sdks/java/extensions/gcp-core/pom.xml [hunk 6]
```diff
@@ -175,6 +171,12 @@
       <scope>provided</scope>
     </dependency>
 
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+      <scope>provided</scope>
```
Comment (dhalperi) [inline sdks/java/extensions/gcp-core/pom.xml]: why this change?

File: sdks/java/io/elasticsearch/pom.xml [hunk 7]
```diff
@@ -86,18 +87,19 @@
...
     <dependency>
       <groupId>com.google.auto.value</groupId>
       <artifactId>auto-value</artifactId>
       <scope>provided</scope>
     </dependency>
 
+    <dependency>
+      <groupId>joda-time</groupId>
+      <artifactId>joda-time</artifactId>
+      <scope>provided</scope>
```
Comment (dhalperi) [inline sdks/java/io/elasticsearch/pom.xml]: why this change?

File: sdks/java/io/jdbc/pom.xml [hunk 8]
```diff
@@ -107,18 +108,19 @@
...
     <dependency>
       <groupId>com.google.auto.value</groupId>
       <artifactId>auto-value</artifactId>
       <scope>provided</scope>
     </dependency>
 
+    <dependency>
+      <groupId>joda-time</groupId>
+      <artifactId>joda-time</artifactId>
+      <scope>provided</scope>
```
Comment (dhalperi) [inline sdks/java/io/jdbc/pom.xml]: ?

Comment (dhalperi) [inline sdks/java/io/jdbc/pom.xml]: just `test`?

File: sdks/java/io/xml/pom.xml [hunk 9]
```diff
@@ -101,7 +101,7 @@
     <dependency>
       <groupId>org.hamcrest</groupId>
       <artifactId>hamcrest-core</artifactId>
-      <version>1.3</version>
+      <scope>test</scope>
```
Comment (dhalperi) [inline sdks/java/io/xml/pom.xml]: scopes are not inherited I believe.

File: examples/java/pom.xml [hunk 10]
```diff
@@ -367,10 +367,6 @@
 
     <plugins>
       <plugin>
-        <artifactId>maven-compiler-plugin</artifactId>
```
Comment (dhalperi) [inline examples/java/pom.xml]: please update Commit description to explain why this is correct. I believe you're saying that compiler-plugin, jar-plugin, and checkstyle are already on by default for all modules so need not be explicitly enabled?

Confirm by analyzing the difference in `mvn help:effective-pom`.

File: sdks/java/io/hadoop/jdk1.8-tests/pom.xml [hunk 11]
```diff
@@ -125,7 +125,6 @@
   </profiles>
 
   <properties>
-    <log4j.core.version>2.6.2</log4j.core.version>
     <hadoop.common.version>2.7.0</hadoop.common.version>
     <guava.version>19.0</guava.version>
```
Comment (dhalperi) [inline sdks/java/io/hadoop/jdk1.8-tests/pom.xml]: looks like opportunities to move this versioning mumbo jumbo into dependency management in the root hadoop pom.xml.

Comment (dhalperi) [inline sdks/java/io/hadoop/jdk1.8-tests/pom.xml]: And add comments for why this is overriding from parent pom.xml.

File: sdks/java/java8tests/pom.xml [hunk 12]
```diff
@@ -47,25 +47,6 @@
...
-      <plugin>
-        <artifactId>maven-compiler-plugin</artifactId>
-      </plugin>
-
-      <plugin>
-        <groupId>org.apache.maven.plugins</groupId>
-        <artifactId>maven-surefire-plugin</artifactId>
-        <configuration>
-          <systemPropertyVariables>
-            <beamUseDummyRunner>false</beamUseDummyRunner>
```
Comment (dhalperi) [inline sdks/java/java8tests/pom.xml]: is this not inherited and used?

confirm by analyzing `mvn help:effective-pom`

File: sdks/java/io/common/pom.xml (hunk 4 above)
Comment (ssisk) [inline sdks/java/io/common/pom.xml]: acknowledged. If we think HashingFn should move into main and/or core, I'd support that (probably not in this PR)

File: sdks/java/io/xml/pom.xml (hunk 9 above)
Comment (aviemzur) [inline sdks/java/io/xml/pom.xml]: If defined in `dependencyManagement` in the parent they should: https://gist.github.com/aviemzur/2a4dc78b542644ea49601f8ff3068b06

File: pom.xml (hunk 5 above)
Comment (iemejia) [inline pom.xml]: ok

File: pom.xml (hunk 1 above)
Comment (iemejia) [inline pom.xml]: ok

File: runners/spark/pom.xml (hunk 2 above)
Comment (iemejia) [inline runners/spark/pom.xml]: Not related will do in separate PR too.

File: sdks/java/extensions/sorter/pom.xml (hunk 3 above)
Comment (iemejia) [inline sdks/java/extensions/sorter/pom.xml]: I wanted to align the version with the one that we are using in the hadoop components and hdfs, but I will remove this since it is unrelated to this PR.

File: sdks/java/extensions/gcp-core/pom.xml (hunk 6 above)
Comment (iemejia) [inline sdks/java/extensions/gcp-core/pom.xml]: I will detail this in a general comment and in the JIRA

File: sdks/java/io/elasticsearch/pom.xml (hunk 7 above)
Comment (iemejia) [inline sdks/java/io/elasticsearch/pom.xml]: see general comment

File: examples/java/pom.xml (hunk 10 above)
Comment (iemejia) [inline examples/java/pom.xml]: ok

File: sdks/java/io/hadoop/jdk1.8-tests/pom.xml (hunk 11 above)
Comment (iemejia) [inline sdks/java/io/hadoop/jdk1.8-tests/pom.xml]: will do in different PR too.

File: sdks/java/java8tests/pom.xml (hunk 12 above)
Comment (iemejia) [inline sdks/java/java8tests/pom.xml]: yes that's the reason to remove it.

**Comment by iemejia** [comment]
I made a mess of this PR, so I am breaking it in multiple small PRs for the issues that don't deal exactly with the JIRA, and I will open a new one later on just for the issues related with the JIRA. I will detail a bit more the case so Dan has some more context for the new review.
